parser/charset/encoding_gbk.go

```diff
@@ -105,3 +107,65 @@ var GBKCase = unicode.SpecialCase{
 	unicode.CaseRange{Lo: 0x01DC, Hi: 0x01DC, Delta: [unicode.MaxCase]rune{0, 0, 0}},
 	unicode.CaseRange{Lo: 0x216A, Hi: 0x216B, Delta: [unicode.MaxCase]rune{0, 0, 0}},
 }
+
+// customGBK is a simplifiedchinese.GBK wrapper.
+type customGBK struct{}
+
+// NewDecoder returns simplifiedchinese.GBK.NewDecoder().
+func (c customGBK) NewDecoder() *encoding.Decoder {
+	return &encoding.Decoder{
+		Transformer: customGBKDecoder{
+			gbkDecoder: simplifiedchinese.GBK.NewDecoder(),
+		},
+	}
+}
+
+type customGBKDecoder struct {
+	gbkDecoder *encoding.Decoder
+}
+
+// Transform special treatment for 0x80,
+// see https://github.com/pingcap/tidb/issues/30581 get details.
+func (c customGBKDecoder) Transform(dst, src []byte, atEOF bool) (nDst, nSrc int, err error) {
+	if len(src) == 0 {
+		return 0, 0, nil
+	}
+	if src[0] == 0x80 {
+		return utf8.EncodeRune(dst[:], utf8.RuneError), 1, nil
+	}
+	return c.gbkDecoder.Transform(dst, src, atEOF)
+}
+
+// Reset is same as simplifiedchinese.GBK.Reset().
+func (c customGBKDecoder) Reset() {
+	c.gbkDecoder.Reset()
+}
+
+type customGBKEncoder struct {
+	gbkEncoder *encoding.Encoder
+}
+
+// NewEncoder returns simplifiedchinese.gbk.
+func (c customGBK) NewEncoder() *encoding.Encoder {
+	return &encoding.Encoder{
+		Transformer: customGBKEncoder{
+			gbkEncoder: simplifiedchinese.GBK.NewEncoder(),
+		},
+	}
+}
+
+// Transform special treatment for `€`,
+// see https://github.com/pingcap/tidb/issues/30581 get details.
+func (c customGBKEncoder) Transform(dst, src []byte, atEOF bool) (nDst, nSrc int, err error) {
+	if len(src) == utf8.RuneLen('€') {
+		if r,_ := utf8.DecodeRune(src[:]); r == '€' {
+			return 0, 0, errInvalidCharacterString
+		}
+	}
```

Comment:
```suggestion
	if bytes.HasPrefix(src, []byte{0xe2, 0x82, 0xac} /* '€' */) {
		return 0, 0, errInvalidCharacterString
	}
```

How about using bytes directly since the utf-8 encoding of '€' should be constant.

Reply:
Here is the result:
```go
func main() {
	var buf [4]byte
	i := utf8.EncodeRune(buf[:], '€')
	euroSign := buf[:i]
	fmt.Printf("%x\n", euroSign)
}
```
```console
e282ac
```